// @tanstack/router-core/dist/cjs/ssr/transformStreamWithRouter.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const web = require("node:stream/web");
const node_stream = require("node:stream");
const utils = require("../utils.cjs");
function transformReadableStreamWithRouter(router, routerStream) {
  return transformStreamWithRouter(router, routerStream);
}
function transformPipeableStreamWithRouter(router, routerStream) {
  return node_stream.Readable.fromWeb(
    transformStreamWithRouter(router, node_stream.Readable.toWeb(routerStream))
  );
}
const TSR_SCRIPT_BARRIER_ID = "$tsr-stream-barrier";
const patternBodyEnd = /(<\/body>)/;
const patternHtmlEnd = /(<\/html>)/;
const patternClosingTag = /(<\/[a-zA-Z][\w:.-]*?>)/g;
function createPassthrough(onCancel) {
  let controller;
  const encoder = new TextEncoder();
  const stream = new web.ReadableStream({
    start(c) {
      controller = c;
    },
    cancel() {
      res.destroyed = true;
      onCancel?.();
    }
  });
  const res = {
    stream,
    write: (chunk) => {
      if (typeof chunk === "string") {
        controller.enqueue(encoder.encode(chunk));
      } else {
        controller.enqueue(chunk);
      }
    },
    end: (chunk) => {
      if (chunk) {
        res.write(chunk);
      }
      controller.close();
      res.destroyed = true;
    },
    destroy: (error) => {
      controller.error(error);
    },
    destroyed: false
  };
  return res;
}
async function readStream(stream, opts) {
  try {
    const reader = stream.getReader();
    let chunk;
    while (!(chunk = await reader.read()).done) {
      opts.onData?.(chunk);
    }
    opts.onEnd?.();
  } catch (error) {
    opts.onError?.(error);
  }
}
function transformStreamWithRouter(router, appStream, opts) {
  let stopListeningToInjectedHtml = void 0;
  let timeoutHandle;
  const finalPassThrough = createPassthrough(() => {
    if (stopListeningToInjectedHtml) {
      stopListeningToInjectedHtml();
      stopListeningToInjectedHtml = void 0;
    }
    clearTimeout(timeoutHandle);
  });
  const textDecoder = new TextDecoder();
  let isAppRendering = true;
  let routerStreamBuffer = "";
  let pendingClosingTags = "";
  let streamBarrierLifted = false;
  let leftover = "";
  let leftoverHtml = "";
  function getBufferedRouterStream() {
    const html = routerStreamBuffer;
    routerStreamBuffer = "";
    return html;
  }
  function decodeChunk(chunk) {
    if (chunk instanceof Uint8Array) {
      return textDecoder.decode(chunk, { stream: true });
    }
    return String(chunk);
  }
  const injectedHtmlDonePromise = utils.createControlledPromise();
  let processingCount = 0;
  handleInjectedHtml();
  stopListeningToInjectedHtml = router.subscribe(
    "onInjectedHtml",
    handleInjectedHtml
  );
  function handleInjectedHtml() {
    router.serverSsr.injectedHtml.forEach((promise) => {
      processingCount++;
      promise.then((html) => {
        if (isAppRendering) {
          routerStreamBuffer += html;
        } else {
          finalPassThrough.write(html);
        }
      }).catch(injectedHtmlDonePromise.reject).finally(() => {
        processingCount--;
        if (!isAppRendering && processingCount === 0) {
          injectedHtmlDonePromise.resolve();
        }
      });
    });
    router.serverSsr.injectedHtml = [];
  }
  injectedHtmlDonePromise.then(() => {
    clearTimeout(timeoutHandle);
    const finalHtml = leftoverHtml + getBufferedRouterStream() + pendingClosingTags;
    finalPassThrough.end(finalHtml);
  }).catch((err) => {
    console.error("Error reading routerStream:", err);
    finalPassThrough.destroy(err);
  }).finally(() => {
    if (stopListeningToInjectedHtml) {
      stopListeningToInjectedHtml();
      stopListeningToInjectedHtml = void 0;
    }
  });
  readStream(appStream, {
    onData: (chunk) => {
      const text = decodeChunk(chunk.value);
      const chunkString = leftover + text;
      const bodyEndMatch = chunkString.match(patternBodyEnd);
      const htmlEndMatch = chunkString.match(patternHtmlEnd);
      if (!streamBarrierLifted) {
        const streamBarrierIdIncluded = chunkString.includes(
          TSR_SCRIPT_BARRIER_ID
        );
        if (streamBarrierIdIncluded) {
          streamBarrierLifted = true;
          router.serverSsr.liftScriptBarrier();
        }
      }
      if (bodyEndMatch && htmlEndMatch && bodyEndMatch.index < htmlEndMatch.index) {
        const bodyEndIndex = bodyEndMatch.index;
        pendingClosingTags = chunkString.slice(bodyEndIndex);
        finalPassThrough.write(
          chunkString.slice(0, bodyEndIndex) + getBufferedRouterStream()
        );
        leftover = "";
        return;
      }
      let result;
      let lastIndex = 0;
      while ((result = patternClosingTag.exec(chunkString)) !== null) {
        lastIndex = result.index + result[0].length;
      }
      if (lastIndex > 0) {
        const processed = chunkString.slice(0, lastIndex) + getBufferedRouterStream() + leftoverHtml;
        finalPassThrough.write(processed);
        leftover = chunkString.slice(lastIndex);
      } else {
        leftover = chunkString;
        leftoverHtml += getBufferedRouterStream();
      }
    },
    onEnd: () => {
      isAppRendering = false;
      router.serverSsr.setRenderFinished();
      if (processingCount === 0) {
        injectedHtmlDonePromise.resolve();
      } else {
        const timeoutMs = opts?.timeoutMs ?? 6e4;
        timeoutHandle = setTimeout(() => {
          injectedHtmlDonePromise.reject(
            new Error("Injected HTML timeout after app render finished")
          );
        }, timeoutMs);
      }
    },
    onError: (error) => {
      console.error("Error reading appStream:", error);
      finalPassThrough.destroy(error);
      injectedHtmlDonePromise.reject(error);
    }
  });
  return finalPassThrough.stream;
}
exports.TSR_SCRIPT_BARRIER_ID = TSR_SCRIPT_BARRIER_ID;
exports.transformPipeableStreamWithRouter = transformPipeableStreamWithRouter;
exports.transformReadableStreamWithRouter = transformReadableStreamWithRouter;
exports.transformStreamWithRouter = transformStreamWithRouter;
